metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractRenameMember.scala

```diff
@@ -0,0 +1,387 @@
+    object traverser extends SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg if p.pos.toLSP.overlapsWith(range) =>
+          packages += Pkg(ref = p.ref, stats = Nil)
+          super.apply(p)
+        case i: Import =>
+          imports += i
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+
+    traverser(tree)
+
+    def names(t: Term): List[Term.Name] = {
+      t match {
+        case s: Term.Select => names(s.qual) :+ s.name
+        case n: Term.Name => n :: Nil
+      }
+    }
+
+    def merge(n1: Term.Ref, n2: Term.Name): Term.Select = n1 match {
+      case s: Term.Select => Term.Select(qual = s, name = n2)
+      case n: Term.Name => Term.Select(qual = n, name = n2)
+    }
+
+    def mergeNames(ns: List[Term.Name]): Option[Term.Ref] = ns match {
+      case Nil => None
+      case head :: Nil => Some(head)
+      case head :: second :: xs => Some(xs.foldLeft(merge(head, second))(merge))
+    }
+
+    val termNames = packages
+      .flatMap(p => names(p.ref))
+
+    val mergedTermsOpt = mergeNames(termNames.toList)
+
+    val pkg: Option[Pkg] = mergedTermsOpt.map(t => Pkg(ref = t, stats = Nil))
+
+    val structure = pkg.toList.mkString("\n") ::
+      imports.mkString("\n") ::
+      member.toString ::
+      companion.map(_.toString).getOrElse("") :: Nil
+
+    structure
+      .filter(_.nonEmpty)
+      .mkString("\n\n")
+  }
+
+  private def parents(member: Member): List[String] = {
+
+    def namesFromTemplate(t: Template): List[String] = {
+      t.inits.flatMap {
+        _.tpe match {
+          case Type.Name(value) => Some(value)
+          case _ => None
+        }
+      }
+    }
+
+    member match {
+      case c: Defn.Class => namesFromTemplate(c.templ)
+      case t: Defn.Trait => namesFromTemplate(t.templ)
+      case o: Defn.Object => namesFromTemplate(o.templ)
+      case e: Defn.Enum => namesFromTemplate(e.templ)
+    }
+  }
+
+  private def renameFileAsMemberAction(
+      uri: String,
+      member: Member
+  ): l.CodeAction = {
+    val className = member.name.value
+    val newUri = newPathFromClass(uri, member).toURI.toString
+    val fileName = uri.toAbsolutePath.filename
+
+    val edits: List[Either[l.TextDocumentEdit, l.ResourceOperation]] = List(
+      Right(new l.RenameFile(uri, newUri))
+    )
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(
+      ExtractRenameMember.renameFileAsClassTitle(fileName, className)
+    )
+    codeAction.setKind(l.CodeActionKind.Refactor)
+    codeAction.setEdit(
+      new l.WorkspaceEdit(edits.map(_.asJava).asJava)
+    )
+
+    codeAction
+  }
+
+  private def extractClassAction(
+      uri: String,
+      member: Member,
+      title: String
+  ): l.CodeAction = {
+
+    val range = member.name.pos.toLSP
+
+    val codeAction = new l.CodeAction()
+    codeAction.setTitle(title)
+    codeAction.setKind(kind)
```

Comment:
```suggestion
    codeAction.setKind(l.CodeActionKind.RefactorExtract)
```
maybe inline it since we have two different action produced in the class.